_drafts/2023-08-15-debugging-android-devices.md

```diff
@@ -0,0 +1,274 @@
+
+[What Terrible Failures](https://developer.android.com/reference/android/util/Log.html#wtf(java.lang.String,%20java.lang.String)) according to Google should be used to:
+
+> Report a condition that should never happen. The error will always be logged at level ASSERT with the call stack. Depending on system configuration, a report may be added to the DropBoxManager, and/or the process may be terminated immediately with an error dialog.
+
+These are logged in Android apps and services throughout the system to record things that as the documentation says, should never happen but we all know that things that should never happen on our systems do still happen, and that's the perfect time to trigger a WTF.
+
+### SeLinux Violations
+
+Android's SELinux (Security-Enhanced Linux) is an implementation of the SELinux framework specifically created by the National Security Agency (NSA). SELinux is used on Android devices to enhance security by enforcing mandatory access controls and ensuring that processes and applications operate within defined security contexts. Android's implementation of SELinux is designed to provide a robust security mechanism to mitigate risks associated with malware, unauthorized access, and other security threats. When incorrectly configured, violations will be logged to logcat via `auditd`. Violations can lead to crashes when the resources attempting to be accessed are protected. One important detail to remember is that in engineering builds, SSLinux is configured to be permissive and will only log a warning but will still allow access to the resource being requested. This can make it easy to miss violations during development.
+
+You can find more details here in the [official documentation](https://source.android.com/docs/security/features/selinux/).
+
+## Diagnostic Tools
+
+Due to the complexity of Android-based devices, Google ships several tools to help diagnose what is happening on them.
+
+### Android Debug Bridge (ADB)
+
+This allows you to connect to your device over USB or the network, run commands, pull logs, install apps, reboot the device, and push/pull files. This is a feature-rich tool that supports several other commands and options but those are the most commonly used.
+
+### Bug Reports
+
+Bug Reports are huge dumps of data from the device that collects information from all the services running on the device and creates a snapshot of the device at the time it was triggered. These are very resource-intensive and often will cause the device to appear frozen so you don’t normally want to do this frequently to a customer’s device.
+
+Android's official documentation has dozens of short explanations on exploring different types of data contained within a Bug Report.
+
+### DropboxManager
+
+DropboxManager is for app specific crash logs. As mentioned above, this is where Java exceptions, tombstones, kernel panics, etc can be found as individual files as opposed to having to try and pull them out of logcat.
```

Comment:
This seems pretty cool. I think fleshing this out with some examples and doc links would be helpful.